File: atd/remind/hooks.py
"""提醒 hook：内置 toast / email，外加用户脚本目录 ~/.atd/hooks/。"""
from __future__ import annotations

import json
import subprocess
import sys
import traceback
from pathlib import Path

from .. import config
from ..model import Task

HOOK_NAMES = ("toast", "email")


def discover_user_hooks() -> list[str]:
    """扫描 ~/.atd/hooks 下的 .py / .bat / .exe / .cmd，名字即 hook 名。"""
    d = config.data_dir() / "hooks"
    out = []
    if d.is_dir():
        for p in sorted(d.iterdir()):
            if p.is_file() and p.suffix.lower() in (".py", ".bat", ".exe", ".cmd", ".ps1"):
                out.append(p.stem)
    return out


def available_hooks() -> list[str]:
    return list(HOOK_NAMES) + discover_user_hooks()


# ---------------------------------------------------------------- 内置 toast
# 三端系统通知：Windows 用 winrt toast，macOS 用 osascript，Linux 用 notify-send。
# toast 名字三端通用，hooks 里写 @18:00:toast 即可。

def fire_toast(task: Task, message: str) -> None:
    if sys.platform == "win32":
        _toast_windows(message)
    elif sys.platform == "darwin":
        _toast_macos(message)
    else:
        _toast_linux(message)


def _toast_windows(message: str) -> None:
    try:
        from windows_toasts import Toast, WindowsToaster
    except ImportError:
        print("atd: 未安装 windows-toasts，跳过 toast", file=sys.stderr)
        return
    t = Toast(text_fields=["atd 提醒", message])
    WindowsToaster("anothertodo").show_toast(t)


def _toast_macos(message: str) -> None:
    """macOS 系统通知（osascript，零依赖）。"""
    # osascript 字符串里的引号/反斜杠需转义，防止脚本注入
    safe = message.replace("\\", "\\\\").replace('"', '\\"')
    script = f'display notification "{safe}" with title "atd 提醒"'
    try:
        subprocess.run(["osascript", "-e", script], check=True,
                       capture_output=True, text=True, timeout=15)
    except (subprocess.SubprocessError, FileNotFoundError, OSError) as e:
        print(f"atd: macOS toast 失败：{e}", file=sys.stderr)


def _toast_linux(message: str) -> None:
    """Linux 桌面通知（notify-send，零依赖）。"""
    try:
        subprocess.run(["notify-send", "atd 提醒", message], check=True,
                       capture_output=True, text=True, timeout=15)
    except (subprocess.SubprocessError, FileNotFoundError, OSError) as e:
        print(f"atd: Linux toast 失败：{e}", file=sys.stderr)


# ---------------------------------------------------------------- 内置 email

def fire_email(task: Task, message: str) -> None:
    import smtplib
    from email.mime.text import MIMEText

    cfg = config.load()["email"]
    if not cfg.get("host") or not cfg.get("to"):
        print("atd: email 未配置（config.toml 的 [email] 段），跳过", file=sys.stderr)
        return
    msg = MIMEText(message, "plain", "utf-8")
    msg["Subject"] = f"[atd] {task.title}"
    msg["From"] = cfg.get("from") or cfg.get("user")
    msg["To"] = cfg["to"]
    cls = smtplib.SMTP_SSL if cfg.get("ssl", True) else smtplib.SMTP
    with cls(cfg["host"], int(cfg.get("port", 465)), timeout=15) as s:
        if cfg.get("user"):
            s.login(cfg["user"], cfg["password"])
        s.sendmail(msg["From"], [cfg["to"]], msg.as_string())


BUILTIN = {"toast": fire_toast, "email": fire_email}


def fire_user_hook(name: str, task: Task, message: str) -> None:
    d = config.data_dir() / "hooks"
    for ext in (".py", ".bat", ".exe", ".cmd", ".ps1"):
        p = d / (name + ext)
        if p.exists():
            payload = json.dumps({"task": task.to_dict(), "message": message}, ensure_ascii=False)
            if p.suffix.lower() == ".py":
                result = subprocess.run([sys.executable, str(p)], input=payload,
                                        capture_output=True, text=True, encoding="utf-8", timeout=60)
            else:
                result = subprocess.run([str(p)], input=payload,
                                        capture_output=True, text=True, encoding="utf-8", timeout=60)
            if result.returncode != 0:
                detail = result.stderr.strip() or result.stdout.strip()
                raise RuntimeError(f"用户 hook {name} 退出码 {result.returncode}: {detail}")
            return
    raise KeyError(f"找不到用户 hook：{name}")


def fire(hook_name: str, task: Task, message: str) -> bool:
    """触发一个 hook，吞掉异常（提醒不能因单个 hook 挂掉而中断）。"""
    try:
        if hook_name in BUILTIN:
            BUILTIN[hook_name](task, message)
        else:
            fire_user_hook(hook_name, task, message)
        return True
    except Exception:
        traceback.print_exc()
        return False


def build_message(task: Task, missed: bool, now_str: str) -> str:
    from ..agenda import format_date
    from datetime import date
    due = format_date(task, date.today()) if task.due else ""
    prefix = "[错过] " if missed else ""
    bits = [prefix + task.title]
    if task.priority:
        bits.append(f"[{task.priority}]")
    if due.strip():
        bits.append(f"日期 {due.strip()}")
    bits.append(f"@{now_str}")
    return "  ".join(bits)

File: src/reminders/hooks.ts
import { readdir } from "node:fs/promises";
import { basename, extname, join } from "node:path";

import { execa } from "execa";
import nodemailer, { type Transporter } from "nodemailer";

import type { Task } from "../contracts.js";
import { dataDir, loadConfig } from "../core/config.js";

export type NotificationInput = { task: Task; message: string };
export interface NotificationHook { readonly name: string; send(input: NotificationInput): Promise<void>; }
export type HookResult = { name: string; ok: boolean; error?: string };
type MailTransportOptions = { host: string; port: number; secure: boolean; auth?: { user: string; pass: string } };
type MailTransportFactory = (options: MailTransportOptions) => Transporter;

type ProcessResult = { exitCode: number; stdout?: string; stderr?: string };
type ProcessRunner = (command: string, args: string[], options: { env: NodeJS.ProcessEnv }) => Promise<ProcessResult>;

const runProcess: ProcessRunner = async (command, args, options) => {
  const result = await execa(command, args, { ...options, reject: false });
  return { exitCode: result.exitCode ?? 1, stdout: result.stdout, stderr: result.stderr };
};

const WINDOWS_TOAST_SCRIPT = `$ErrorActionPreference = 'Stop'
Add-Type -AssemblyName System.Runtime.WindowsRuntime
$shortcutSource = @'
using System;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.ComTypes;
using System.Text;

public static class AtdToastShortcut {
    [ComImport, Guid("00021401-0000-0000-C000-000000000046")]
    private class ShellLink { }
    [ComImport, Guid("000214F9-0000-0000-C000-000000000046"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    private interface IShellLinkW {
        int GetPath(StringBuilder path, int max, IntPtr findData, uint flags);
        int GetIDList(out IntPtr idList); int SetIDList(IntPtr idList);
        int GetDescription(StringBuilder description, int max); int SetDescription(string description);
        int GetWorkingDirectory(StringBuilder directory, int max); int SetWorkingDirectory(string directory);
        int GetArguments(StringBuilder arguments, int max); int SetArguments(string arguments);
        int GetHotkey(out short hotkey); int SetHotkey(short hotkey);
        int GetShowCmd(out int showCmd); int SetShowCmd(int showCmd);
        int GetIconLocation(StringBuilder path, int max, out int index); int SetIconLocation(string path, int index);
        int SetRelativePath(string path, uint reserved); int Resolve(IntPtr hwnd, uint flags); int SetPath(string path);
    }
    [ComImport, Guid("886D8EEB-8CF2-4446-8D02-CDBA1DBDCF99"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    private interface IPropertyStore {
        int GetCount(out uint count); int GetAt(uint index, out PropertyKey key); int GetValue(ref PropertyKey key, out PropVariant value);
        int SetValue(ref PropertyKey key, ref PropVariant value); int Commit();
    }
    [StructLayout(LayoutKind.Sequential, Pack = 4)] private struct PropertyKey { public Guid FormatId; public uint PropertyId; }
    [StructLayout(LayoutKind.Explicit)] private struct PropVariant {
        [FieldOffset(0)] public ushort VariantType;
        [FieldOffset(8)] public IntPtr Pointer;
    }
    public static void Install(string shortcutPath, string target, string arguments, string appId) {
        var link = (IShellLinkW)new ShellLink();
        Marshal.ThrowExceptionForHR(link.SetPath(target));
        Marshal.ThrowExceptionForHR(link.SetArguments(arguments));
        var properties = (IPropertyStore)link;
        var key = new PropertyKey { FormatId = new Guid("9F4C2855-9F79-4B39-A8D0-E1D42DE1D5F3"), PropertyId = 5 };
        var value = new PropVariant { VariantType = 31, Pointer = Marshal.StringToCoTaskMemUni(appId) };
        try { Marshal.ThrowExceptionForHR(properties.SetValue(ref key, ref value)); Marshal.ThrowExceptionForHR(properties.Commit()); }
        finally { Marshal.FreeCoTaskMem(value.Pointer); }
        ((IPersistFile)link).Save(shortcutPath, true);
    }
}
'@
if (-not ('AtdToastShortcut' -as [type])) { Add-Type -TypeDefinition $shortcutSource }
$appId = 'anothertodo.atd'
$shortcutDir = Join-Path $env:APPDATA 'Microsoft\\Windows\\Start Menu\\Programs'
New-Item -ItemType Directory -Force -Path $shortcutDir | Out-Null
$shortcutPath = Join-Path $shortcutDir 'anothertodo.lnk'
$powershellPath = (Get-Command powershell.exe).Source
[AtdToastShortcut]::Install($shortcutPath, $powershellPath, '-NoProfile -WindowStyle Hidden', $appId)
$payload = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String($env:ATD_TOAST_PAYLOAD)) | ConvertFrom-Json
$escape = { param([string]$value) [System.Security.SecurityElement]::Escape($value) }
$title = & $escape ([string]$payload.title)
$message = & $escape ([string]$payload.message)
$xmlText = "<toast><visual><binding template='ToastGeneric'><text>$title</text><text>$message</text></binding></visual></toast>"
$xmlType = [System.Type]::GetType('Windows.Data.Xml.Dom.XmlDocument, Windows, ContentType=WindowsRuntime')
if ($null -eq $xmlType) { throw 'Windows.Data.Xml.Dom.XmlDocument WinRT type unavailable' }
$xml = [System.Activator]::CreateInstance($xmlType)
$xml.LoadXml($xmlText)
$toastType = [System.Type]::GetType('Windows.UI.Notifications.ToastNotification, Windows, ContentType=WindowsRuntime')
if ($null -eq $toastType) { throw 'Windows.UI.Notifications.ToastNotification WinRT type unavailable' }
$toast = [System.Activator]::CreateInstance($toastType, [object[]]@($xml))
$managerType = [System.Type]::GetType('Windows.UI.Notifications.ToastNotificationManager, Windows, ContentType=WindowsRuntime')
if ($null -eq $managerType) { throw 'Windows.UI.Notifications.ToastNotificationManager WinRT type unavailable' }
$method = $managerType.GetMethod('CreateToastNotifier', [System.Type[]]@([string]))
if ($null -eq $method) { throw 'CreateToastNotifier WinRT method unavailable' }
$notifier = $method.Invoke($null, [object[]]@($appId))
$notifier.Show($toast)
`;

export const windowsToastArgs = (): string[] => ["-NoProfile", "-NonInteractive", "-WindowStyle", "Hidden", "-ExecutionPolicy", "Bypass", "-Command", WINDOWS_TOAST_SCRIPT];
export const windowsToastEnvironment = (message: string): NodeJS.ProcessEnv => ({
  ...process.env,
  ATD_TOAST_PAYLOAD: Buffer.from(JSON.stringify({ title: "atd 提醒", message }), "utf8").toString("base64"),
});

export const sendWindowsToast = async (message: string, runner: ProcessRunner = runProcess): Promise<void> => {
  const result = await runner("powershell", windowsToastArgs(), { env: windowsToastEnvironment(message) });
  if (result.exitCode !== 0) throw new Error(`Windows toast 失败：${result.stderr || result.stdout || `exit ${result.exitCode}`}`);
};

const toastWindows: NotificationHook = {
  name: "toast",
  async send({ message }) { await sendWindowsToast(message); },
};
const toastMacos: NotificationHook = {
  name: "toast",
  async send({ message }) { const result = await execa("osascript", ["-e", `display notification ${JSON.stringify(message)} with title "atd 提醒"`], { reject: false }); if (result.exitCode !== 0) throw new Error(`macOS toast 失败：${result.stderr || result.stdout || `exit ${result.exitCode}`}`); },
};
const toastLinux: NotificationHook = {
  name: "toast",
  async send({ message }) { const result = await execa("notify-send", ["atd 提醒", message], { reject: false }); if (result.exitCode !== 0) throw new Error(result.stderr || "notify-send 失败"); },
};

export const toastHook = (): NotificationHook => process.platform === "win32" ? toastWindows : process.platform === "darwin" ? toastMacos : toastLinux;

export const createEmailHook = (dir = dataDir(), transportFactory: MailTransportFactory = (options) => nodemailer.createTransport(options as never)): NotificationHook => ({
  name: "email",
  async send({ task, message }) {
    const config = await loadConfig(dir);
    const password = process.env.ATD_EMAIL_PASSWORD ?? config.email.password;
    if (!config.email.host || !config.email.to) throw new Error("email hook 未配置 host/to");
    const transportOptions: MailTransportOptions = { host: config.email.host, port: config.email.port, secure: config.email.ssl, ...(config.email.user ? { auth: { user: config.email.user, pass: password } } : {}) };
    const transporter = transportFactory(transportOptions);
    await transporter.sendMail({ from: config.email.from || config.email.user, to: config.email.to, subject: `atd 提醒：${task.title}`, text: message });
  },
});
export const emailHook: NotificationHook = createEmailHook();

const supportedExtensions = new Set([".js", ".cmd", ".bat", ".ps1", ".py", ".exe"]);
export const discoverUserHooks = async (dir = dataDir()): Promise<string[]> => {
  try { return (await readdir(join(dir, "hooks"))).filter((name) => supportedExtensions.has(extname(name).toLowerCase())).map((name) => basename(name, extname(name))); }
  catch { return []; }
};

export const runUserHook = async (name: string, input: NotificationInput, dir = dataDir(), timeout = 15_000): Promise<void> => {
  const hookDir = join(dir, "hooks");
  const names = (await discoverUserHooks(dir)).filter((candidate) => candidate === name);
  if (!names.length) throw new Error(`找不到用户 hook：${name}`);
  const files = await readdir(hookDir);
  const file = files.find((candidate) => basename(candidate, extname(candidate)) === name && supportedExtensions.has(extname(candidate).toLowerCase()));
  if (!file) throw new Error(`找不到用户 hook：${name}`);
  const path = join(hookDir, file);
  const ext = extname(file).toLowerCase();
  const command = ext === ".py" ? "python" : ext === ".js" ? process.execPath : ext === ".ps1" ? "powershell" : path;
  const args = ext === ".py" || ext === ".js" ? [path] : ext === ".ps1" ? ["-NoProfile", "-NonInteractive", "-File", path] : [];
  const result = await execa(command, args, { input: JSON.stringify(input), timeout, reject: false });
  if (result.timedOut) throw new Error(`用户 hook 超时：${name}`);
  if (result.exitCode !== 0) throw new Error(`用户 hook 失败：${name} ${result.stderr || result.stdout}`);
};

export const hookNames = async (dir = dataDir()): Promise<string[]> => ["toast", "email", ...(await discoverUserHooks(dir))];

export const fireHook = async (name: string, input: NotificationInput, dir = dataDir()): Promise<HookResult> => {
  try {
    if (name === "toast") await toastHook().send(input);
    else if (name === "email") await createEmailHook(dir).send(input);
    else await runUserHook(name, input, dir);
    return { name, ok: true };
  } catch (error) {
    return { name, ok: false, error: error instanceof Error ? error.message : String(error) };
  }
};
